### ISSUE TITLE
Issues with scrollbar in Eclipse (Mint 20)

### ISSUE THREAD

**Issue by MiniDVLover** [B00merang-Project/Windows-XP#29]
Both versions 3.0 and 3.1 of the XP theme have a scrollbar graphics bug. The scrollbars are, for the most part, normal, except in Eclipse (see the attached photo). It seems to occur on all Eclipse windows with scrollbars, and if I switch to a different theme, even another version of Windows, the scrollbars are displayed normally. (if I switch to a different version of XP, the bug remains)

Additionally, in any LibreOffice program, the horizontal scrollbar "backdrop" seems to be a duplicated version of the vertical graphic; it goes back to normal with any other theme, as with Eclipse.

Any idea what's causing this? It started after I upgraded from Mint 19.3 to 20 (with Cinnamon 4.6.7 desktop).

![Scrollbar Bug](https://user-images.githubusercontent.com/58407173/96663313-1657e780-1316-11eb-8581-5d1c49c081ac.png)

**Comment by Elbullazul** [comment]
if you launch eclipse from the terminal, does it show any warnings or errors??

**Comment by MiniDVLover** [comment]
Yes- interesting:

`aidan@Aidan-PC:~$ /home/aidan/Downloads/eclipse/eclipse
org.eclipse.m2e.logback.configuration: The org.eclipse.m2e.logback.configuration bundle was activated before the state location was initialized.  Will retry after the state location is initialized.
org.eclipse.m2e.logback.configuration: Logback config file: /media/aidan/B270D88170D84E2B/Users/Aiden/eclipse-workspace-Java/.metadata/.plugins/org.eclipse.m2e.logback.configuration/logback.1.16.0.20200318-1040.xml
org.eclipse.m2e.logback.configuration: Initializing logback

(Eclipse:59553): Gtk-CRITICAL **: 15:15:19.705: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:19.705: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.069: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.408: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.411: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.411: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.412: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.412: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.426: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-CRITICAL **: 15:15:20.635: gtk_box_gadget_distribute: assertion 'size >= 0' failed in GtkScrollbar

(Eclipse:59553): Gtk-WARNING **: 15:15:20.640: Negative content width -8 (allocation 2, extents 5x5) while allocating gadget (node trough, owner GtkProgressBar)

(Eclipse:59553): Gtk-WARNING **: 15:15:20.640: Negative content width -8 (allocation 2, extents 5x5) while allocating gadget (node trough, owner GtkProgressBar)

(Eclipse:59553): Gtk-WARNING **: 15:15:20.640: Negative content width -8 (allocation 2, extents 5x5) while allocating gadget (node trough, owner GtkProgressBar)

(Eclipse:59553): Gtk-WARNING **: 15:15:20.640: Negative content width -8 (allocation 2, extents 5x5) while allocating gadget (node trough, owner GtkProgressBar)

`

**Comment by Elbullazul** [comment]
In Help > About > Installation Details > Configuration, do you see a GTK version, kind of like this?

> org.eclipse.swt.internal.gtk.version=3.14.12

**Comment by MiniDVLover** [comment]
Something like this?
![Eclipse Version](https://user-images.githubusercontent.com/58407173/96926343-132a3c00-147b-11eb-973e-5b8ebe5e8ab9.png)

**Comment by MiniDVLover** [comment]
Oops - here is is with a wider version column.
![Eclipse Version 2](https://user-images.githubusercontent.com/58407173/96926553-508ec980-147b-11eb-9cbd-d8697b012dc2.png)

**Comment by Elbullazul** [comment]
the info I'm looking for isn't here.

can you tell me if this command works? If not, can you post the result?

> GTK_DEBUG=interactive eclipse

**Comment by MiniDVLover** [comment]
In the terminal, correct?

The command result was:
<code>
aidan@Aidan-PC:~$ GTK_DEBUG=interactive eclipse
eclipse: command not found
</code>

**Comment by Elbullazul** [comment]
how did you install eclipse? I'm going to have to test it by myself

**Comment by MiniDVLover** [comment]
I downloaded a .tar.gz file from here: https://www.eclipse.org/downloads/download.php?file=/technology/epp/downloads/release/2020-09/R/eclipse-java-2020-09-R-linux-gtk-x86_64.tar.gz

I extracted the "eclipse" folder into my "downloads" folder. Opening that and clicking "eclipse" starts the program (no formal installation required).

**Comment by Elbullazul** [comment]
on 19 it looks ok on my side:
![Linux Mint 19 2 Cinnamon-2020-10-25-12-38-08](https://user-images.githubusercontent.com/15310985/97113239-5e349100-16bf-11eb-89f4-6e165f0a77ad.png)


I'm upgrading to 20 and see if it pops

**Comment by Elbullazul** [comment]
managed to reproduce it, it seems like the code originally written for GTK 3.20 is now incompatible. However, the windows 7 theme isn't affected.

I'll have to adapt Windows 7's scrollbars to reproduce the look of Windows XP. It may take a while, but once one is completed, the rest should come easily

**Comment by MiniDVLover** [comment]
Sounds good- thanks!

**Comment by Elbullazul** [comment]
sorry for the delay, but I keep trying to fix more minor bugs on other themes. I should be addressing this issue in the coming days

**Comment by MiniDVLover** [comment]
No worries. Thanks!

The same bug seems to occur in the latest version of Angry IP Scanner as well (which is also in Java).

**Comment by Elbullazul** [comment]
So, I found the cause of the bug. When loading a Java app, the CSS parser includes some rules from Adwaita and other integrated themes, and one of those rules is to add some padding to scrollbar buttons. Unfortunately, there's nothing I can do on my side to fix this, as I don't know how to prevent that behaviour from happening. I originally planned to port the scrollbar portion to a working code base, but found that it affected other themes as well, and some of those have features incompatible with the new approach.

I also noticed that themes with scrollbar buttons disabled are not affected, but those who have them enabled will be affected by this bug. So, for example, the macOS theme will not be affected.

There is a fix, however. By creating the following file and adding the code snippet, it will override the ruleset from root themes and fix the issue:

The file is named `gtk.css` and has to be located in `~/.config/gtk-3.0/gtk.css`.

Inside, paste this snippet of code:
```css
scrollbar button { padding: 0px; }
```

Result:
![image](https://user-images.githubusercontent.com/15310985/101268576-65c16e00-3732-11eb-9861-af46b2b0392c.png)

Remind to rename or remove the file if it affects other themes.
